### PR TITLE
AST transform for denormalising type references

### PR DESCRIPTION
**Pull request by ColinEberhardt**
When compiling a module, this AST transform denormalises type references so that func / call_indirect instructions have inline signatures.

As a result, the interpreter now supports the following:

```
(module
  (type $i32return (func (result i32)))
  (func (type $i32return) (i32.const 1))
)
```

### CODE REVIEW

**Comment by maurobringolf** [comment]
Really cool!

Do I read this correctly that you dont remove the type instruction from the ast? Reading diffs on mobile is hard 😄

**Comment by ColinEberhardt** [comment]
> Do I read this correctly that you dont remove the type instruction from the ast?
Correct -  leave them as-is.

File: packages/ast/test/transform/denormalize-type-references/index.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  }
+  if (a.trim() !== b.trim()) {
+    throw new Error("Assertion error");
+  }
+};
+
+const loadJSON = filename =>
+  JSON.parse(readFileSync(path.join(__dirname, filename), "utf8"));
+
+const copy = json => JSON.parse(JSON.stringify(json));
```
Comment (xtuc) [inline packages/ast/test/transform/denormalize-type-references/index.js]: I think that we can use `const copy = o => Object.assign({}, o);` here.

Comment (ColinEberhardt) [reply packages/ast/test/transform/denormalize-type-references/index.js]: I think that will just perform a shallow copy?

Comment (xtuc) [reply packages/ast/test/transform/denormalize-type-references/index.js]: That's right. I was a bit worried about perf but it's in the test so doesn't matter so much :+1:

**Pull request merged**